drbg/src/hashdrbg.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{
+    helpers::{slice_to_u896, truncate_from_start},
+    props::{
+        MAX_BYTES_ADDITIONAL_INPUT, MAX_BYTES_PERSONALIZATION_STRING, MAX_BYTES_PER_REQUEST,
+        MAX_REQUESTS_BETWEEN_RESEEDS,
+    },
+    HashDrbgProps,
+};
+use core::iter::once;
+use crypto_bigint::{prelude::ArrayEncoding, U896};
+pub use digest;
+use digest::{generic_array::GenericArray, typenum::Unsigned, Digest, Output};
+use tpm2_rs_server::platform::crypto::{
+    drbg_helpers::{next_u32_via_fill, next_u64_via_fill},
+    Drbg, DrbgError,
+};
+
+/// [HashDrbg] is a generic software-based construct for creating
+/// a Hash DRBG compliant with
+/// [NIST SP 800-90Ar1](http://dx.doi.org/10.6028/NIST.SP.800-90Ar1).
+///
+/// It is based on the specifications outlined in _10.1 DRBG Mechanisms Based
+/// on Hash Functions_. This DRBG mechanism implementation uses the
+/// highest security strength of available in the hash function. Thus
+/// `requested_security_strength` parameter is omitted from arguments to
+/// `instantiate` function as well as from the internal state.
+///
+/// ## Notes
+///  `Hash` is expected to be cryptographically secure hash function.
+///
+pub struct HashDrbg<Hash: Digest + HashDrbgProps> {
```

Comment:
Would it make sense to introduce a trait for DRBG with init/generate/reseed/drop functions to allow different implementation, potentially using HW acceleration? This implementation would be one of many.
Also, for HW it would be great to make sure that there is no assumption that state can be copied.

Reply:
Actually we have two layer of traits here and you can pick which one you prefer, The first is software using Hardware `Hash` in software `HashDrbg` which just requires setting the right parameters.

The otherway is Fully hardware accelerated Drbg, in which case you don't really need this crate at all, and all you would need is the traits in the platform api from [server/src/platform/crypto/drbg/mod.rs](https://github.com/tpm-rs/tpm-rs/blob/main/server/src/platform/crypto/drbg/mod.rs)